File: apps/api/src/jobs/mmfRateUpdater.ts
import cron from 'node-cron';
import { MmfFund } from '../models/MmfFund';

/**
 * Scheduled job to update MMF rates (every 6 hours)
 * In production, this would fetch from a data provider API
 */
export function startMMFRateUpdater(): void {
  cron.schedule('0 */6 * * *', async () => {
    console.log('📈 Updating MMF rates...');

    try {
      // Simulate slight rate fluctuations using MongoDB
      const funds = await MmfFund.find({ isActive: true });

      for (const fund of funds) {
        const currentRate = fund.interestRate;
        // Random fluctuation of ±0.1%
        const fluctuation = (Math.random() - 0.5) * 0.2;
        const newRate = Math.max(5, Math.min(25, currentRate + fluctuation));

        fund.interestRate = parseFloat(newRate.toFixed(3));
        await fund.save();
      }

      console.log(`✅ Updated rates for ${funds.length} MMFs in MongoDB`);
    } catch (error) {
      console.error('❌ MMF rate update error:', error);
    }
  });

  console.log('⏰ MMF rate updater scheduled (every 6 hours)');
}

File: OneDrive/Desktop/SharedProj/HazinaHub/backend/app/jobs/mmf_rate_updater.py
import random
from datetime import datetime
from apscheduler.schedulers.asyncio import AsyncIOScheduler

from app.models.mmf_fund import MmfFund


async def run_mmf_rate_updater():
    """Simulate MMF rate fluctuations every 6 hours."""
    print("[INFO] Updating MMF rates...")
    try:
        funds = await MmfFund.find(MmfFund.is_active == True).to_list()
        for fund in funds:
            fluctuation = (random.random() - 0.5) * 0.2
            new_rate = max(5.0, min(25.0, fund.interest_rate + fluctuation))
            fund.interest_rate = round(new_rate, 3)
            fund.updated_at = datetime.utcnow()
            await fund.save()
        print(f"[OK] Updated rates for {len(funds)} MMFs in MongoDB")
    except Exception as e:
        print(f"[ERROR] MMF rate update error: {e}")


def register_mmf_rate_updater(scheduler: AsyncIOScheduler):
    scheduler.add_job(run_mmf_rate_updater, "cron", hour="*/6", id="mmf_rate_updater")
    print("[SCHEDULED] MMF rate updater scheduled (every 6 hours)")
